src/testing/snaptest.zig

```diff
@@ -0,0 +1,220 @@
+        defer std.testing.allocator.free(got);
+
+        try snapshot.diff(got);
+    }
+
+    // Compare the snapshot with the json serialization of a `value`.
+    pub fn diff_json(snapshot: *const Snap, value: anytype) !void {
+        var got = std.ArrayList(u8).init(std.testing.allocator);
+        defer got.deinit();
+
+        try std.json.stringify(value, .{}, got.writer());
+        try snapshot.diff(got.items);
+    }
+
+    // Compare the snapshot with a given string.
+    pub fn diff(snapshot: *const Snap, got: []const u8) !void {
+        if (std.mem.eql(u8, got, snapshot.text)) return;
+
+        std.debug.print(
+            \\Snapshot differs.
+            \\Want:
+            \\----
+            \\{s}
+            \\----
+            \\Got:
+            \\----
+            \\{s}
+            \\----
+            \\
+        ,
+            .{
+                snapshot.text,
+                got,
+            },
+        );
+
+        if (!snapshot.should_update()) {
+            return error.SnapDiff;
+        }
+
+        var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
+        defer arena.deinit();
+
+        const allocator = arena.allocator();
+
+        const file_text =
+            try std.fs.cwd().readFileAlloc(allocator, snapshot.location.file, 1024 * 1024);
+        var file_text_updated = try std.ArrayList(u8).initCapacity(allocator, file_text.len);
+
+        const line_zero_based = snapshot.location.line - 1;
+        const range = snap_range(file_text, line_zero_based) orelse
+            return error.SnapNotFound;
+
+        const snapshot_prefix = file_text[0..range.start];
+        const snapshot_text = file_text[range.start..range.end];
+        const snapshot_suffix = file_text[range.end..];
+
+        const indent = get_indent(snapshot_text);
+
+        try file_text_updated.appendSlice(snapshot_prefix);
+        {
+            var lines = std.mem.split(u8, got, "\n");
+            while (lines.next()) |line| {
+                try file_text_updated.writer().print("{s}\\\\{s}\n", .{ indent, line });
+            }
+        }
+        try file_text_updated.appendSlice(snapshot_suffix);
+
+        try std.fs.cwd().writeFile(snapshot.location.file, file_text_updated.items);
+
+        std.debug.print("Updated {s}\n", .{snapshot.location.file});
+        return error.SnapUpdated;
+    }
+};
+
+const Range = struct { start: usize, end: usize };
+fn snap_range(text: []const u8, snap_line: u32) ?Range {
+    var offset: usize = 0;
+    var line_number: u32 = 0;
+
+    var lines = std.mem.split(u8, text, "\n");
+    const snap_start = while (lines.next()) |line| : (line_number += 1) {
+        if (line_number == snap_line) {
+            assert(std.mem.indexOf(u8, line, "@src()") != null);
+        }
+        if (line_number == snap_line + 1) {
+            assert(is_multiline_string(line));
+            break offset;
+        }
+        offset += line.len + 1; // 1 for \n
+    } else return null;
```

Comment:
Should this be `else unreachable`? How could this be hit?

Reply:
Yeah, I am inconsistent here. This _could_ be fixed if the file is modified in-place before the test is run. In general, there's fundamental fragility to update workflow. But it should be fine in practice, and it probably is better to assert consistently here.